File: ui/src/test/fixtures.ts
/**
 * The daemon's rows, as a test needs one of them.
 *
 * A `TaskDto` is fifteen fields and a test cares about two, so every test file
 * was writing out the other thirteen — and picking its own ids, its own
 * timestamps and its own defaults doing it. Each builder here answers a
 * complete, valid row, and takes an overlay of whatever the test is actually
 * about: `aTask({ status: "approved" })`.
 *
 * One timestamp for everything, because a test that cares about time says so by
 * overriding it, and one id per kind, because a test that needs two rows apart
 * says that by overriding the id too.
 */

import type {
  AgentConfigDto,
  AgentEventDto,
  components,
  EffortDto,
  GoalDto,
  ModelDto,
  RepositoryDto,
  SessionDto,
  SessionEntryDto,
  SkillDto,
  TaskDto,
} from "@/api"
import type { OutsideSessionDto } from "@/features/sessions/queries"

type SessionPageDto = components["schemas"]["SessionPageDto"]

/** The instant everything the daemon holds was created and last touched. */
const STAMP = "2026-01-01T00:00:00Z"

const GOAL_ID = "01JGOAL0000000000000000001"
const TASK_ID = "01JTASK0000000000000000001"
const SESSION_ID = "01JSESS0000000000000000001"
const AUTHOR_ID = "01JAGENT0000000000000AUTH"
const REVIEWER_ID = "01JAGENT0000000000000REVW"
const REPO_ID = "01JREPO0000000000000000001"

/** A row nobody has reported tokens for, which is how every fixture starts. */
const NO_TOKENS = { input_tokens: 0, cached_input_tokens: 0, output_tokens: 0 }

export function aGoal(overrides: Partial<GoalDto> = {}): GoalDto {
  return {
    id: GOAL_ID,
    title: "Ship the board",
    description: "",
    model: "claude-agent-acp:claude-sonnet-5",
    orchestrated: true,
    repos: [],
    status: "active",
    usage: {
      total: NO_TOKENS,
      orchestrator: NO_TOKENS,
      authors: NO_TOKENS,
      reviewers: NO_TOKENS,
    },
    created_at: STAMP,
    updated_at: STAMP,
    ...overrides,
  }
}

export function aTask(overrides: Partial<TaskDto> = {}): TaskDto {
  return {
    id: TASK_ID,
    goal_id: GOAL_ID,
    title: "Wire the sessions screen",
    description: "",
    status: "in_progress",
    branch: "wire-the-sessions-screen-000001",
    landing: "merge",
    repo_id: REPO_ID,
    stalled: false,
    agents: [
      {
        id: AUTHOR_ID,
        seat: "author",
        skills: ["coding"],
        model: "claude-agent-acp:claude-sonnet-5",
      },
      {
        id: REVIEWER_ID,
        seat: "reviewer",
        skills: ["code-review"],
        model: "claude-agent-acp:claude-sonnet-5",
      },
    ],
    depends_on: [],
    picks: [],
    usage: { total: NO_TOKENS, author: NO_TOKENS, reviewers: [] },
    created_at: STAMP,
    updated_at: STAMP,
    ...overrides,
  }
}

export function aSession(overrides: Partial<SessionDto> = {}): SessionDto {
  const id = overrides.id ?? SESSION_ID
  return {
    id,
    goal_id: GOAL_ID,
    task_id: TASK_ID,
    seat: "author",
    task_agent_id: AUTHOR_ID,
    model: "claude-agent-acp:claude-sonnet-5",
    internal_session_id: null,
    worktree_path: null,
    status: "running",
    attention_reason: null,
    attention_since: null,
    last_activity_at: STAMP,
    usage: NO_TOKENS,
    context_used: null,
    context_size: null,
    created_at: STAMP,
    ended_at: null,
    ...overrides,
  }
}

/**
 * A stored session an ACP agent holds on its own, listed over
 * `GET /v1/sessions?kind=outside` — Ariadne did not start it, and it carries
 * no goal, task or seat, having none of its own.
 */
export function anOutsideSession(overrides: Partial<OutsideSessionDto> = {}): OutsideSessionDto {
  return {
    agent_id: "claude-agent-acp",
    internal_session_id: "outside-internal-id-1",
    working_directory: "/Users/me/dev/ariadne",
    last_activity_at: STAMP,
    first_prompt: "Fix the flaky test.",
    ...overrides,
  }
}

/**
 * One page of `GET /v1/sessions` holding these Ariadne sessions, as the
 * daemon answers it: every one a listing row of kind `ariadne`.
 */
export function aSessionPage(
  sessions: SessionDto[],
  page: Partial<SessionPageDto> = {},
): SessionPageDto {
  return aPage(
    sessions.map(
      ({ worktree_path, ...session }): SessionEntryDto => ({
        ...session,
        kind: "ariadne",
        agent_id: session.model.split(":")[0] ?? "",
        title: null,
        working_directory: worktree_path,
      }),
    ),
    page,
  )
}

/**
 * One page of `GET /v1/sessions?kind=outside` holding these conversations,
 * as the daemon answers it.
 */
export function anOutsideSessionPage(
  sessions: OutsideSessionDto[],
  page: Partial<SessionPageDto> = {},
): SessionPageDto {
  return aPage(
    sessions.map(
      (session): SessionEntryDto => ({
        kind: "outside",
        id: session.internal_session_id,
        agent_id: session.agent_id,
        internal_session_id: session.internal_session_id,
        working_directory: session.working_directory,
        last_activity_at: session.last_activity_at,
        title: session.first_prompt,
      }),
    ),
    page,
  )
}

function aPage(sessions: SessionEntryDto[], page: Partial<SessionPageDto>): SessionPageDto {
  return {
    sessions,
    next_cursor: null,
    total: sessions.length,
    snapshot_at: STAMP,
    ...page,
  }
}

export function anAgentEvent(overrides: Partial<AgentEventDto> = {}): AgentEventDto {
  return {
    id: "01JEVENT000000000000000001",
    session_id: SESSION_ID,
    task_id: TASK_ID,
    kind: "post_tool_use",
    summary: "Read AGENTS.md.",
    payload: { tool_name: "Read", cwd: "/Users/me/dev/ariadne" },
    created_at: STAMP,
    ...overrides,
  }
}

export function aSkill(overrides: Partial<SkillDto> = {}): SkillDto {
  return {
    name: "coding",
    seat: "task",
    summary: "Implement a task from its specification.",
    document: "---\nname: coding\ndescription: Implement a task from its specification.\n---\n",
    document_is_default: true,
    builtin: true,
    created_at: STAMP,
    updated_at: STAMP,
    ...overrides,
  }
}

export function aRepository(overrides: Partial<RepositoryDto> = {}): RepositoryDto {
  return {
    id: REPO_ID,
    path: "/home/me/dev/ariadne",
    base_branch: "main",
    description: "The orchestrator itself.",
    created_at: STAMP,
    updated_at: STAMP,
    ...overrides,
  }
}

export function anAgentConfig(overrides: Partial<AgentConfigDto> = {}): AgentConfigDto {
  return {
    agent_id: "claude-agent-acp",
    extra_flags: [],
    default_flags: [],
    ...overrides,
  }
}

/** One reasoning effort a `GET /v1/models` entry can be run at. */
export function anEffort(overrides: Partial<EffortDto> = {}): EffortDto {
  return {
    id: "medium",
    description: null,
    default: false,
    ...overrides,
  }
}

/** One entry of the model catalog `GET /v1/models` serves. */
export function aModel(overrides: Partial<ModelDto> = {}): ModelDto {
  return {
    id: "claude-agent-acp:claude-sonnet-5",
    agent_id: "claude-agent-acp",
    description: null,
    efforts: [],
    enabled: true,
    rank: null,
    ...overrides,
  }
}

File: crates/ariadne-cli/src/commands/fixtures.rs
//! One plausible row of each DTO, for the unit tests.
//!
//! Every test that renders a goal, a task, a session, a profile or a
//! repository needs a whole DTO to render, and only ever cares about two or
//! three of its fields. These build the rest once; the caller names what it is
//! testing with struct-update syntax:
//!
//! ```ignore
//! let stalled = TaskDto { stalled: true, ..fixtures::task("01T", "01G") };
//! ```

use ariadne_api::goals::GoalDto;
use ariadne_api::profiles::ProfileDto;
use ariadne_api::repositories::RepositoryDto;
use ariadne_api::sessions::SessionDto;
use ariadne_api::tasks::TaskDto;
use ariadne_core::{AgentKind, GoalStatus, MergeStrategy, Seat, SessionStatus, TaskStatus};

/// A stamp every fixture is created and updated at, so a rendered row is
/// reproducible.
pub const NOW: &str = "2026-08-18T10:00:00Z";

pub fn goal(id: &str, title: &str) -> GoalDto {
    GoalDto {
        id: id.into(),
        title: title.into(),
        description: String::new(),
        status: GoalStatus::Active,
        max_tasks: None,
        required_approvals: 1,
        orchestrator_profile_id: "01PROFILE".into(),
        model: None,
        effort: None,
        repos: Vec::new(),
        usage: Default::default(),
        created_at: NOW.into(),
        updated_at: NOW.into(),
    }
}

/// A task in progress, titled after its own id.
pub fn task(id: &str, goal_id: &str) -> TaskDto {
    TaskDto {
        id: id.into(),
        goal_id: goal_id.into(),
        repo_id: "01REPO".into(),
        title: format!("task {id}"),
        description: String::new(),
        status: TaskStatus::InProgress,
        author_profile_id: "01ENG".into(),
        author_profile_name: Some("Author".into()),
        orchestrator_profile_name: Some("Orchestrator".into()),
        model: None,
        effort: None,
        reviewers: Vec::new(),
        depends_on: Vec::new(),
        branch: format!("a-task-{id}"),
        worktree_path: None,
        review_round: 0,
        stalled: false,
        merge_commit: None,
        pr_url: None,
        reason: None,
        usage: Default::default(),
        created_at: NOW.into(),
        updated_at: NOW.into(),
    }
}

/// A running session: an author's when it names a task, an orchestrator's
/// when it does not, and one the daemon has raised no attention flag for.
pub fn session(id: &str, goal_id: &str, task_id: Option<&str>) -> SessionDto {
    SessionDto {
        id: id.into(),
        goal_id: goal_id.into(),
        task_id: task_id.map(Into::into),
        seat: match task_id {
            Some(_) => Seat::Author,
            None => Seat::Orchestrator,
        },
        profile_id: "01PROF".into(),
        agent_kind: AgentKind::ClaudeCode,
        model: None,
        effort: None,
        internal_session_id: None,
        tmux_session: format!("ariadne-{id}"),
        worktree_path: None,
        review_round: None,
        status: SessionStatus::Running,
        attention_reason: None,
        attention_since: None,
        last_activity_at: None,
        usage: Default::default(),
        created_at: NOW.into(),
        ended_at: None,
    }
}

/// A profile with a system prompt of its own, on no particular agent.
pub fn profile(name: &str, seat: Seat) -> ProfileDto {
    ProfileDto {
        id: format!("01{name}"),
        name: name.into(),
        seat,
        model: None,
        effort: None,
        system_prompt: "you are an author".into(),
        system_prompt_is_default: false,
        created_at: "2026-08-17T08:00:00Z".into(),
        updated_at: "2026-08-17T09:00:00Z".into(),
    }
}

pub fn repository(id: &str, path: &str, base_branch: &str) -> RepositoryDto {
    RepositoryDto {
        id: id.into(),
        path: path.into(),
        base_branch: base_branch.into(),
        merge_strategy: MergeStrategy::Direct,
        description: None,
        landing_prompt: "Land {branch} onto {base_branch}.".into(),
        landing_prompt_is_default: true,
        created_at: NOW.into(),
        updated_at: NOW.into(),
    }
}
